src/ucs/config/parser.c

```diff
@@ -1169,51 +1169,63 @@ static ucs_status_t ucs_config_apply_env_vars(void *opts, ucs_config_field_t *fi
     return UCS_OK;
 }
 
-static const char * ucs_config_parser_get_base_prefix(const char *env_prefix)
+/* Find if env_prefix consists of base prefix and sub prefix and returns sub prefix
+ * length, base prefix should not contain underscores "_" except at the very end
+ * e.g in env_prefix = "LONG_PREFIX_UCX_":
+ * "LONG_PREFIX_" is sub prefix
+ * "UCX_"         is base prefix
+ */
+static ucs_status_t ucs_config_parser_get_sub_prefix_len(const char *env_prefix, 
+                                                         size_t *sub_prefix_len)
 {
-    int sub_prefix_len;
-    
+    size_t len;
+
     /* env_prefix always has "_" at the end and we want to find the last but one
      * "_" in the env_prefix */
-    sub_prefix_len = strlen(env_prefix) - 2;
+    len = strlen(env_prefix);
+    if (len < 2 ) {
+        ucs_error("Invalid value of env_prefix");
+        return UCS_ERR_INVALID_PARAM;
+    }
 
-    while ((sub_prefix_len >= 0) && (env_prefix[sub_prefix_len] != '_')) {
-        sub_prefix_len -= 1;
+    len -= 2;
+    while ((len > 0) && (env_prefix[len - 1] != '_')) {
+        len -= 1;
     }
-    return env_prefix + sub_prefix_len;
+    *sub_prefix_len = len;
+
+    return UCS_OK;
 }
 
 ucs_status_t ucs_config_parser_fill_opts(void *opts, ucs_config_field_t *fields,
                                          const char *env_prefix,
                                          const char *table_prefix,
                                          int ignore_errors)
 {
+    size_t       sub_prefix_len = 0;
     ucs_status_t status;
-    const char   *base_prefix;
 
     /* Set default values */
     status = ucs_config_parser_set_default_values(opts, fields);
-    ucs_assert(env_prefix != NULL);
     if (status != UCS_OK) {
         goto err;
     }
 
-    /* Find if env_prefix consists of base prefix and sub prefix, base prefix
-     * should not contain underscores "_" except at the very end
-     * e.g in env_prefix = "LONG_PREFIX_UCX_":
-     * "LONG_PREFIX_" is sub prefix
-     * "UCX_"         is base prefix */
-    base_prefix = ucs_config_parser_get_base_prefix(env_prefix);
+    ucs_assert(env_prefix != NULL);
+    status = ucs_config_parser_get_sub_prefix_len(env_prefix, &sub_prefix_len);
```

Comment:
the function can return **pointer** to sub-prefix, or NULL if it doesn't exist, something like this:
```
status = ucs_config_parser_get_sub_prefix_len(env_prefix, &sub_prefix);
...

if (sub_prefix != NULL) {
   ucs_config_apply_env_vars(..., sub_prefix); /* would apply UCX_xx vars */
}

ucs_config_apply_env_vars(..., env_prefix); /* would apply MPI_UCX_xx vars */
```